### PR TITLE
more comments and more tests

### PR DESCRIPTION
**Pull request by natebrennand**
This PR adds more tests for the `packet` package and provides more comments in the binary marshaling functions. It also adds several examples from the gearman protocol page.

### CODE REVIEW

File: packet/packet.go [hunk 1]
```diff
@@ -51,27 +69,38 @@ func (packet *Packet) UnmarshalBinary(data []byte) error {
 
 // MarshalBinary implements the encoding.BinaryMarshaler interface
 func (packet *Packet) MarshalBinary() ([]byte, error) {
+	// form a buffer with the packet's magic code
 	buf := bytes.NewBuffer(packet.Code)
+
+	// write the request header
```
Comment (azylman) [inline packet/packet.go]: Technically the magic code, the type, and the size all comprise the header, so this comment is a bit inaccurate.

File: packet/packet.go [hunk 2]
```diff
@@ -51,27 +69,38 @@ func (packet *Packet) UnmarshalBinary(data []byte) error {
...
 	}
-	if len(packet.Arguments) > 0 {
-		// Need special handling for last argument (don't write null byte)
-		for _, argument := range packet.Arguments[0 : len(packet.Arguments)-1] {
-			buf.Write(argument)
-			buf.Write([]byte{0})
+
+	// write all arguments provided
+	for i := 0; i < len(packet.Arguments); i++ {
+		if _, err := buf.Write(packet.Arguments[i]); err != nil {
```
Comment (azylman) [inline packet/packet.go]: The API for `buffer.Write` specifies that the error is always nil, so you don't need to check this response

File: packet/packet.go [hunk 3]
```diff
@@ -51,27 +69,38 @@ func (packet *Packet) UnmarshalBinary(data []byte) error {
...
+			return nil, fmt.Errorf("Error while writing packet argument: %s", err)
+		}
+
+		// null deliminate every argument but the last
```
Comment (azylman) [inline packet/packet.go]: I almost think this was easier to read the other way, which went through each argument but the last and then wrote the last with different logic

File: packet/packet_test.go [hunk 4]
```diff
@@ -1,42 +1,116 @@
...
+			0x48, 0x3a, 0x6c, 0x61, 0x70, 0x3a, 0x31, 0x00,
+			0x72, 0x65, 0x76, 0x65, 0x72, 0x73, 0x65, 0x00,
+			0x74, 0x65, 0x73, 0x74,
+		},
 	},
 	{
-		"pack":  Packet{Code: Req, Type: 3, Arguments: [][]byte{}},
-		"bytes": []byte{0, 0x52, 0x45, 0x51, 0, 0, 0, 3, 0, 0, 0, 0},
+		/*
+			00 52 45 51                \0REQ        (Magic)
```
Comment (azylman) [inline packet/packet_test.go]: love these changes

**Comment by azylman** [comment]
couple nits but lgtm

File: packet/packet.go (hunk 2 above)
Comment (natebrennand) [inline packet/packet.go]: TIL

**Pull request merged**